File: binding-csharp/runtime/src/main/csharp/Org.Apache.Etch.Bindings.Csharp/Transport/Fmt/TypeCode.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
namespace Org.Apache.Etch.Bindings.Csharp.Transport.Fmt
{
    /// <summary>
    /// Values denoting types of encoded values. There are two value ranges:
    /// 
    ///  -128 to -65: codes for defined types.
    /// -64 to 127: tiny integer values.
    ///            
    /// Tiny integers are completely encoded within the type byte.
    /// </summary>
    public class TypeCode
    {
        /// <summary>
        /// A code denoting a null value.
        /// </summary>
        public const sbyte NULL = -128;

        /// <summary>
        /// A code denoting no value. 
        /// For example, an array is a sequence of values (some of which may
	    /// be NULL), terminated by a NONE.
        /// </summary>
        public const sbyte NONE = -127;


        /// <summary>
        /// A code denoting a false boolean value.
        /// </summary>
        public const sbyte BOOLEAN_FALSE = -126;

        /// <summary>
        /// A code denoting a true boolean value.
        /// </summary>
        public const sbyte BOOLEAN_TRUE = -125;

        /// <summary>
        /// A code denoting a signed const sbyte.
        /// </summary>
        public const sbyte BYTE = -124;

        /// <summary>
        /// A code denoting a single const sbyte signed short.
        /// </summary>
        public const sbyte SHORT = -123;

        /// <summary>
        /// A code denoting a four sbyte signed integer.
        /// </summary>
        public const sbyte INT = -122;
             
        /// <summary>
        /// A code denoting an eight const sbyte signed long, lsb first.
        /// </summary>
        public const sbyte LONG = -121;

        /// <summary>
        /// A code denoting a four const sbyte ieee floating format number.
        /// </summary>
        public const sbyte FLOAT = -120;

        /// <summary>
        /// A code denoting an eight const sbyte ieee floating format number.
        /// </summary>
        public const sbyte DOUBLE = -119;

        //// Arrays /////

        /// <summary>
        /// A code denoting an array of booleans.
        /// </summary>
        //public const sbyte BOOLS = -118;

        /// <summary>
        /// A code denoting an array of bytes.
        /// </summary>
        public const sbyte BYTES = -117;

        /// <summary>
        /// A code denoting an array of shorts.
        /// </summary>
        //public const sbyte SHORTS = -116;

        /// <summary>
        /// A code denoting an array of ints.
        /// </summary>
        //public const sbyte INTS = -115;

        /// <summary>
        /// A code denoting an array of longs.
        /// </summary>
        //public const sbyte LONGS = -114;

        /// <summary>
        /// A code denoting an array of singles.
        /// </summary>
        //public const sbyte FLOATS = -113;

        /// <summary>
        /// A code denoting an array of doubles.
        /// </summary>
        //public const sbyte DOUBLES = -112;

        /// <summary>
        /// A code denoting a sequence of values.
        /// </summary>
        public const sbyte ARRAY = -111;


        //// strings /////
        
        /// <summary>
        /// A code denoting an empty string.
        /// </summary>
        public const sbyte EMPTY_STRING = -110;

        /// <summary>
        /// A code denoting a utf-8 encoded string.
        /// </summary>
        public const sbyte STRING = -109;

        /// <summary>
        /// A code denoting a custom value (struct, exception, enum, extern)
	    /// from a value factory. An associated value identifies the specific
	    /// type. The format on the wire of STRUCT and CUSTOM are the same.
	    /// Emit CUSTOM instead of STRUCT, and accept STRUCT as if it were
	    /// CUSTOM.
        /// </summary>
        public const sbyte STRUCT = -108;



        /// <summary>
        /// A code denoting a custom value (struct, exception, enum, extern)
        /// from a value factory. An associated value identifies the specific
        /// type. The format on the wire of STRUCT and CUSTOM are the same.
        /// Emit CUSTOM instead of STRUCT, and accept STRUCT as if it were
        /// CUSTOM.
        /// </summary>
        public const sbyte CUSTOM = -107;


        /// <summary>
        /// A code denoting that any value is ok (as long as we know how
        /// to serialize it). Dynamic methods should be applied to determine
        /// the type. This type never appears on the wire.
        /// </summary>
        public const sbyte ANY = -106;

        ////////////////////
        // tiny integers //
        ////////////////////

        /// <summary>
        /// Minimum "small" integer.
        /// </summary>
        public const sbyte MIN_TINY_INT = -64;

        /// <summary>
        /// Maximum "small" integer. Small integers are encoded asis
        /// (with embedded type code)
        /// </summary>
        public const sbyte MAX_TINY_INT = 127;

        
    }
}

File: binding-ruby/src/main/ruby/transport/fmt/type_code.rb
module TypeCode
    
#  /**
#   * A code denoting a null value.
#   */
  NULL = -1;
#  /**
#   * A code denoting a false boolean value.
#   */
  BOOLEAN_FALSE = -2;
#  /**
#   * A code denoting a true boolean value.
#   */
  BOOLEAN_TRUE = -3;
#  /**
#   * A code denoting a signed byte.
#   */
  BYTE1 = -4;
#  /**
#   * A code denoting a single byte signed short.
#   */
  SHORT1 = -5;
#  /**
#   * A code denoting a two byte signed short, lsb first.
#   */
  SHORT2 = -6;
#  /**
#   * A code denoting a single byte signed integer.
#   */
  INT1 = -7;
#  /**
#   * A code denoting a two byte signed integer, lsb first.
#   */
  INT2 = -8;
#  /**
#   * A code denoting a four byte signed integer, lsb first.
#   */
  INT4 = -9;
#  /**
#   * A code denoting a single byte signed long.
#   */
  LONG1 = -10;
#  /**
#   * A code denoting a two byte signed long, lsb first.
#   */
  LONG2 = -11;
#  /**
#   * A code denoting a four byte signed long, lsb first.
#   */
  LONG4 = -12;
#  /**
#   * A code denoting an eight byte signed long, lsb first.
#   */
  LONG8 = -13;
#  /**
#   * A code denoting a four byte ieee floating format number.
#   */
  FLOAT4 = -14;
#  /**
#   * A code denoting an eight byte ieee floating format number.
#   */
  FLOAT8 = -15;
#  /**
#   * A code denoting an array of bytes.
#   */
  BYTES = -16;
#  /**
#   * A code denoting an empty string.
#   */
  EMPTY_STRING = -17;
#  /**
#   * A code denoting a utf-8 encoded string.
#   */
  STRING = -18;
#  /**
#   * A code denoting a sequence of key / value pairs.
#   */
  STRUCT = -19;
#  /**
#   * A code denoting a sequence of values.
#   */
  ARRAY = -20;
#  /**
#   * A code denoting a custom value from a value factory. An integer
#   * value follows which identifies the specific type.
#   */
  CUSTOM = -21;
#  /**
#   * A code denoting no value, which is different than NULL. For
#   * example, an array is a sequence of values (some of which may
#   * be NULL), terminated by a NONE.
#   */
  NONE = -22;

#  ////////////////////
#  // SMALL INTEGERS //
#  ////////////////////
#  
#  /**
#   * Minimum "small" integer.
#   */
  MIN_SMALL_INT = -16;
#  /**
#   * Maximum "small" integer. Small integers are encoded asis
#   * (with embedded type code)
#   */
  MAX_SMALL_INT = 31;
#  /**
#   * Positive sentinal mask.
#   */
  PSMASK = 0x80;
#  /**
#   * Positive sentinal value.
#   */
  PSVALUE = 0x00;
#  /**
#   * Shift for positive data type value.
#   */
  PDTSHIFT = 5;
#  /**
#   * Mask for positive value.
#   */
  PVMASK = 31;
#  /**
#   * Negative sentinal mask.
#   */
  NSMASK = 0xC0;
#  /**
#   * Negative sentinal value.
#   */
  NSVALUE = 0x80;
#  /**
#   * Shift for negative data type value.
#   */
  NDTSHIFT = 4;
#  /**
#   * Mask for negative value.
#   */
  NVMASK = 15;
#  /**
#   * Mask for data type value.
#   */
  DTMASK = 0x03;
#  /**
#   * Data type value for small integer from byte.
#   */
  BYTE_DT = 0;
#  /**
#   * Data type value for small integer from short.
#   */
  SHORT_DT = 1;
#  /**
#   * Data type value for small integer from int.
#   */
  INT_DT = 2;
#  /**
#   * Data type value for small integer from long.
#   */
  LONG_DT = 3;    
end